### PR TITLE
Fix polyline color issue involving duplicate positions

### LINKED ISSUE #9379: Polyline issue with per-vertex colors and duplicate position values
There's a problem when using per-vertex coloring of polylines when the polyline point values have duplicate position values.

Sandcastle example:
https://sandcastle.cesium.com/index.html#c=fVNhb9owEP0rFp9SiTllCE3LKBqiKavESsUQXdVUk0kOsObY0dkJY1X/++ykgZIhIlnyu3vvdHd+KRiSgsMWkFwRCVsyAs3zlC7KmBe14hKPlDSMS8CodfElkpEsrA5tZKm2x8J7JXbCMsegUjC4814iSUimNDdcSR3UvBFDY29MdukKVXoNawTQQ0S285yCkKcPn3pt8tEe3yfxhsk1ELPhmmwA7U2RMt/tuastZSpVHW3X8PMx7HQauNvAvRI/X5SRLU/MJiCdS3pZ4gJs139uFKbMBM2RR0ooHGYZMGQyBroIZ/Pw56+b6ez7cN521V2J2LH0PeCirBWQFRMa2odUQJ7qHTlMp7Ph3TisG6y+I8IsvD6TfQwnk+nDGcLocXh3Jv3w7XYePlf9CZZmczVGlcskIAZziOTrwRAZ8tQ+cwENS9ThygvrN2fcSm3cnuzA78jjRrbSHFRBbTsXfq1eie13HpyyYuNdvBfbOZNa5DFI87Z/W+owSul9qm0a6H4mTVmSeHvoqK12q6/NTsCgXt9XnmYKDclReJT6BtJMMGtOf5nHv8HQWGsndNS+/17aT3hBeHJ14odza9faZla5ED/4X4hag75v+f9JhWIJl+updalgO0fbdAaTKkgp7fsWnlYapcSSYaPyPw

The sandcastle demo shows the problem nicely. Per-vertex coloring can (I think, at least) only be specified when using `PolylineGeometry` and then adding a `Primitive`. The demo adds some polygon points, all different, and all segments with a specific color. For extra visibility, the first point is placed "below" the other points:

grafik

When changing the first polygon point to be the same as the second, the second point is removed by CesiumJS and the colors "shift" a segment:

grafik

The reason for this "shift" is the call to `arrayRemoveDuplicates()`, here:
https://github.com/CesiumGS/cesium/blob/f05522e405137591c591f73dd660bd1ba5d625a7/Source/Core/PolylineGeometry.js#L312

This removes duplicate positions, but doesn't handle removing the per-vertex color indices. A fix would be to pass additional arrays to `arrayRemoveDuplicates()` where entries would also be removed for every position entry removed.

A workaround would be to use `arrayRemoveDuplicates()` (which currently is an undocumented API call) before generating the per-vertex color values. In my case this would work, since I derive the color values from climb or sink values of the track.

Browser: Firefox current (85.0.2)

Operating System: Windows 10

### PR DESCRIPTION
**Pull request by j9liu**
Fixes #9379. This PR changes how colors are handled in `PolylineGeometry.createGeometry`:
- `arrayRemoveDuplicates` now accepts an optional `removedIndices` parameter. If any entries are removed from the original array during this function, the indices of those entries will be stored in `removedIndices` in sorted order.
- `createGeometry` will use this array to remove the colors that correspond to those extraneous positions. This should work for both segment coloring and per-vertex coloring.

Additionally, I noticed that if `wrapAround` were set to `true` in the parameters, and if an array's first and last entries were equal, `arrayRemoveDuplicates` would remove the first entry. However, the examples in the documentation suggested that the last entry was supposed to be removed instead. I changed the function to match the example.

-------------------------------------
### Segment Coloring
 [This sandcastle](https://sandcastle.cesium.com/index.html#c=fVNhb9owEP0rFp9SiTllCE3LKBqiKavESsUQXdVUk0kOsObY0dkJY1X/++ykgZIhIlnyu3vvdHd+KRiSgsMWkFwRCVsyAs3zlC7KmBe14hKPlDSMS8CodfElkpEsrA5tZKm2x8J7JXbCMsegUjC4814iSUimNDdcSR3UvBFDY29MdukKVXoNawTQQ0S285yCkKcPn3pt8tEe3yfxhsk1ELPhmmwA7U2RMt/tuastZSpVHW3X8PMx7HQauNvAvRI/X5SRLU/MJiCdS3pZ4gJs139uFKbMBM2RR0ooHGYZMGQyBroIZ/Pw56+b6ez7cN521V2J2LH0PeCirBWQFRMa2odUQJ7qHTlMp7Ph3TisG6y+I8IsvD6TfQwnk+nDGcLocXh3Jv3w7XYePlf9CZZmczVGlcskIAZziOTrwRAZ8tQ+cwENS9ThygvrN2fcSm3cnuzA78jjRrbSHFRBbTsXfq1eie13HpyyYuNdvBfbOZNa5DFI87Z/W+owSul9qm0a6H4mTVmSeHvoqK12q6/NTsCgXt9XnmYKDclReJT6BtJMMGtOf5nHv8HQWGsndNS+/17aT3hBeHJ14odza9faZla5ED/4X4hag75v+f9JhWIJl+updalgO0fbdAaTKkgp7fsWnlYapcSSYaPyPw) was linked in the issue to demonstrate the bug. It sets up the following polyline:

image

In the current version of Cesium, if the orange segment is collapsed, the remaining segments will be colored incorrectly as shown.

image

The changes in PR result in this polyline:

image

-------------------------------------

### Vertex Coloring
We can modify the sandcastle to have per-vertex coloring [like so](https://sandcastle.cesium.com/index.html#c=fVNhb9owEP0rFp9SiTllCE3LUrQsTVk1WiqK6KqmmkxygDXHjmwnjFX977OTBkqGiBTJ7+690935uSQSlRQ2INEF4rBBIShaZHhexZy4k1Q4FFwTykHGnbMvMY95aXTSRBZicyi8E2zLDHMEIgMtt85LzBHKhaKaCq68hhcSqc2J8D5eSpFdwkoCqEBKsnWsAqGnD58GXfTR/K6LkjXhK0B6TRVagzQngap8f2CPppSuVU2028DPh7DXa+F+Cw8q/HxWRTY01WsP9c7xeYVLMF3/uRIyI9prjxwKJmSQ50Ak4QngeTSdRT9/XU2mN8Gsa6vbEollqTuQ86qWh7QsoLvPeOipWZHFeDINbkdR01/9HRC+jYPwx4n8NLo8kX2MxuPJwwlC+Bjcnkg/fL+eRc91/4xk+UyMpCh4Ws8V89e9X3JJM+OCElqOacK1VVZvxrnmSts1moW8I49a2VqzV3mNK234tb5EsrsS75hTW9fmvJjOCVesSICbS14SpsCU2o9SPQ2sTBrwbiaFSZo6O2ipnW7HV3rLYNis7yvNciE1KiRzMHY1ZDkjxrvuokh+g8aJUlZoqb77XuqntEQ0vTjyHu3alTKZZcHYPf0LcWfou4b/n5QJklK+mhgTM7K1tHVvOK6DGGPfNfC4UgvBFkS2Kv8D).

image

If we try to remove the black vertex by making it a duplicate, in the current version of Cesium, we'll find that it just shifts over to the other vertices:

image

However, we can remove it with these PR's changes:

image

cc @lilleyse for review

EDIT: now that I'm thinking about it, in the per-vertex example it probably makes more sense for the orange point to be removed rather than the middle black vertex, because that preserves the original coloring. I'll work a fix tomorrow

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @j9liu!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/polyline-color-bug/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

File: Source/Core/PolylineGeometry.js [hunk 1]
```diff
@@ -309,10 +309,45 @@ PolylineGeometry.createGeometry = function (polylineGeometry) {
...
+    false,
+    removedIndices
   );
+
+  if (defined(colors) && removedIndices.length > 0) {
+    var removedArrayIndex = 0;
+    var length = colors.length;
+    var nextRemovedIndex = removedIndices[0];
+    var colorsScratch = [];
+    var ii;
```
Comment (lilleyse) [inline Source/Core/PolylineGeometry.js:325]: It's fine to reuse `i` rather than create a new variable

File: Source/Core/PolylineGeometry.js [hunk 2]
```diff
@@ -309,10 +309,45 @@ PolylineGeometry.createGeometry = function (polylineGeometry) {
...
+          removedArrayIndex++;
+          nextRemovedIndex = removedIndices[removedArrayIndex];
+        } else {
+          colorsScratch.push(colors[ii]);
+        }
+      }
+    }
+
+    colors = colorsScratch;
+  }
```
Comment (lilleyse) [inline Source/Core/PolylineGeometry.js:349]: Is it possible to use [`array.filter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) to simplify the code here?

**Review by lilleyse — COMMENTED** [review]
Thanks for the detailed summary at the top of PR. It was good to see that you tested a bunch of different cases. And nice catch with the `wrapAround` bug.

> EDIT: now that I'm thinking about it, in the per-vertex example it probably makes more sense for the orange point to be removed rather than the middle black vertex, because that preserves the original coloring. I'll work a fix tomorrow

Thank for pointing this out. I'll look out for your fix tomorrow.

**Comment by j9liu** [comment]
Context for just-committed changes: I ran into a case today where if I have an array `[A, B, C, ...., A, A, A]`, the function deletes the last two `A` entries, and then it will also remove the final duplicate `A` when it sees `wrapAround` is true. But it doesn't properly account for the index of that `A` with respect to the original array; its index is `length - 3`, but the code will only push `values.length - 1`. Even if I used `cleanValues.length - 1`, the indices will be out of order because it takes care of this case at the very end. 

This commit should take care of this case, but the `wrapAround` step is a check that happens at the end of the loop. I'm not sure how to cleanly integrate it into the existing `for` loop. Will work on `PolylineGeometry.createGeometry` next.

**Comment by j9liu** [comment]
Fix just pushed. Now the sandcastle should look like this.

image

I handled endpoint cases differently than cases where the point is in the middle of the polyline. If a polyline looks like `[A, B, B, C]` with the colors `[red, green, blue, white]`, it arbitrarily the color of the point's first appearance, so the line's colors will become `[red, green, white]`. But when the point is on the end, then it makes more sense to have it drop the first-appearing color.

**Comment by j9liu** [comment]
@lilleyse - made the changes you suggested; let me know if there's anything else that needs adjustment!

**Comment by lilleyse** [comment]
Thanks, the updates look great!

**Pull request merged**